### PR TITLE
feat: calculate dB difference

### PR DESCRIPTION
**Pull request by tormodAase**

### CODE REVIEW

**Review by dansvend — COMMENTED** [review]
Great work🤩 I left some questions.

File: utils/calculateDecibelDifference.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    return createError("VOLUME_RESULTS_MISSING");
+  }
+  if (withoutPlugs === 0 || withPlugs === 0) {
+    return createError("ZERO_VALUES_NOT_ALLOWED");
+  }
+  if (withPlugs < withoutPlugs) {
+    return createError("VOLUME_HIGHER_WITHOUT_PLUGS");
+  }
+  const rawValue = 20 * Math.log10(withoutPlugs / withPlugs);
+  return createValue(parseFloat(rawValue.toFixed(2)));
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply utils/calculateDecibelDifference.ts:25]: I think we should note it down and talk to the product owners before deciding the amount of decimals

File: __tests__/getDecibelDifferenceResult.test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      },
+      right: {
+        withoutPlugs: 0.1,
+        withPlugs: 1,
+      },
+    };
+
+    const expectedValue1: DecibelDifferenceResult = {
+      left: createError("VOLUME_HIGHER_WITHOUT_PLUGS"),
+      right: createValue(-20),
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply __tests__/getDecibelDifferenceResult.test.ts]: fixed

File: __tests__/getDecibelDifferenceResult.test.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      },
+      right: {
+        withoutPlugs: 0.1,
+        withPlugs: 1000,
+      },
+    };
+
+    const expectedValue2: DecibelDifferenceResult = {
+      left: createError("VOLUME_HIGHER_WITHOUT_PLUGS"),
+      right: createValue(-80),
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply __tests__/getDecibelDifferenceResult.test.ts]: Fixed

File: __tests__/calculateDecibelDifference.test.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { createError, createValue, Error } from "../utils/valueOrError";
+describe("calculateDecibelDifference", () => {
+  it("Should calculate the decibel difference for volume X and Y", () => {
+    expect(calculateDecibelDifference(0.1, 1)).toEqual(createValue(-20));
+    expect(calculateDecibelDifference(0.02, 0.5)).toEqual(createValue(-27.96));
+    expect(calculateDecibelDifference(0.06, 0.4)).toEqual(createValue(-16.48));
+    expect(calculateDecibelDifference(0.06, 0.42)).toEqual(createValue(-16.9));
+  });
+
+  it("Should return an error object if the arguments are invalis", () => {
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply __tests__/calculateDecibelDifference.test.ts]: Fixed

File: __tests__/calculateDecibelDifference.test.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  CalculateDecibelDifferenceError,
+  calculateDecibelDifference,
+} from "../utils/calculateDecibelDifference";
+import { createError, createValue, Error } from "../utils/valueOrError";
+describe("calculateDecibelDifference", () => {
+  it("Should calculate the decibel difference for volume X and Y", () => {
+    expect(calculateDecibelDifference(0.1, 1)).toEqual(createValue(-20));
+    expect(calculateDecibelDifference(0.02, 0.5)).toEqual(createValue(-27.96));
+    expect(calculateDecibelDifference(0.06, 0.4)).toEqual(createValue(-16.48));
+    expect(calculateDecibelDifference(0.06, 0.42)).toEqual(createValue(-16.9));
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply __tests__/calculateDecibelDifference.test.ts]: Fixed

File: contexts/_internal/useDecibelDifference.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { useMemo } from "react";
+import { getDecibelDifferenceResult } from "./getDecibelDifferenceResult";
+import { DecibelDifferenceResult, EarVolumeResults } from "./types";
+export const useDecibelDifference = (
+  results: EarVolumeResults,
+): DecibelDifferenceResult => {
+  return useMemo(() => {
+    return getDecibelDifferenceResult(results);
+  }, [results]);
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply contexts/_internal/useDecibelDifference.ts]: Fixed

File: contexts/_internal/types.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+export type EarVolumeResults = Record<Ear, EarVolumeResult>;
+export type EarResultKey = keyof EarVolumeResult;
+export type EarVolumeResult = {
+  withPlugs: number | null;
+  withoutPlugs: number | null;
+};
+export type DecibelDifferenceResult = {
+  left: CalculateDecibelDifferenceResult;
+  right: CalculateDecibelDifferenceResult;
+};
```
[the remark this answers is not in the document]
Comment (tormodAase) [reply contexts/_internal/types.ts]: Fixed

**Review by dansvend — APPROVED** [review]

**Pull request merged**